File: src/ipc/person.ts
import { invoke } from "@tauri-apps/api/core";

export type Person = {
	id: number;
	name: string;
	email: string | null;
	phone: string | null;
	linkedSpeakerCount: number;
};

type Raw = {
	id: number;
	name: string;
	email: string | null;
	phone: string | null;
	linked_speaker_count: number;
};

export type PersonInput = {
	name: string;
	email?: string | null;
	phone?: string | null;
};

const fromRaw = (r: Raw): Person => ({
	id: r.id,
	name: r.name,
	email: r.email,
	phone: r.phone,
	linkedSpeakerCount: r.linked_speaker_count,
});

export const personList = async (): Promise<Person[]> =>
	(await invoke<Raw[]>("person_list")).map(fromRaw);

export const personCreate = async (input: PersonInput): Promise<Person> =>
	fromRaw(
		await invoke<Raw>("person_create", {
			name: input.name,
			email: input.email ?? null,
			phone: input.phone ?? null,
		}),
	);

export const personRename = (
	personId: number,
	input: PersonInput,
): Promise<void> =>
	invoke("person_rename", {
		personId,
		name: input.name,
		email: input.email ?? null,
		phone: input.phone ?? null,
	});

export const personDelete = (personId: number): Promise<void> =>
	invoke("person_delete", { personId });

File: src-tauri/src/commands/person.rs
use crate::commands::util::project_conn;
use crate::db::queries::person::{self, Person};
use crate::error::AppResult;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PersonDTO {
    pub id: i64,
    pub name: String,
    pub linked_speaker_count: i64,
}

impl From<Person> for PersonDTO {
    fn from(value: Person) -> Self {
        Self {
            id: value.id,
            name: value.name,
            linked_speaker_count: value.linked_speaker_count,
        }
    }
}

#[tauri::command]
pub async fn person_list(app: tauri::AppHandle) -> AppResult<Vec<PersonDTO>> {
    let conn = project_conn(&app)?;
    Ok(person::list(&conn)?.into_iter().map(Into::into).collect())
}

#[tauri::command]
pub async fn person_create(app: tauri::AppHandle, name: String) -> AppResult<PersonDTO> {
    let conn = project_conn(&app)?;
    Ok(person::create(&conn, &name)?.into())
}

#[tauri::command]
pub async fn person_rename(
    app: tauri::AppHandle,
    person_id: i64,
    name: String,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    person::rename(&conn, person_id, &name)
}

#[tauri::command]
pub async fn person_delete(app: tauri::AppHandle, person_id: i64) -> AppResult<()> {
    let mut conn = project_conn(&app)?;
    person::delete(&mut conn, person_id)
}
